src/backend/distributed/utils/metadata_cache.c

```diff
@@ -911,6 +968,119 @@ InitializeDistTableCache(void)
 
 
 /*
+ * GetWorkerNodeHash is a wrapper around InitializeWorkerNodeCache(). It
+ * triggers InitializeWorkerNodeCache when the workerHash is NULL. Otherwise,
+ * it returns the hash.
+ */
+HTAB *
+GetWorkerNodeHash(void)
+{
+	if (!workerNodeHashValid)
+	{
+		InitializeWorkerNodeCache();
+
+		workerNodeHashValid = true;
+	}
+
+	return WorkerNodeHash;
+}
+
+
+/*
+ * Initialize the infrastructure for the worker node cache. The functions
+ * reads the worker nodes from the metadata table, adds them to the hash and
+ * finally registers an invalidation callaback.
+ */
+static void
+InitializeWorkerNodeCache(void)
+{
+	static bool invalidationRegistered = false;
+	HTAB *oldWorkerNodeHash = NULL;
+	HTAB *newWorkerNodeHash = NULL;
+	List *workerNodeList = NIL;
+	ListCell *workerNodeCell = NULL;
+	HASHCTL info;
+	int hashFlags = 0;
+	long maxTableSize = (long) MaxWorkerNodesTracked;
+
+	/* make sure we've initialized CacheMemoryContext */
+	if (CacheMemoryContext == NULL)
+	{
+		CreateCacheMemoryContext();
+	}
+
+	/*
+	 * Create the hash that holds the worker nodes. The key is the unique nodeid
+	 * field.
+	 */
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(uint32);
```

Comment:
I think we should address it here. We probably won't do #801 before the code freeze and after looking at the planner code again, it's not entirely clear to me whether #801 provides performance benefit over being able to do a hash look up by nodename/nodeport, since WorkerNodeActive gets called very frequently.